pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/gates/tests/Test_TNCuda_NonParam.cpp

```diff
@@ -0,0 +1,859 @@
+    }
+}
+
+TEMPLATE_LIST_TEST_CASE("TNCuda::applyControlledOperation non-param "
+                        "one-qubit with controls",
+                        "[MPSTNCuda]", TestTNBackends) {
+
+    using TNDevice_T = TestType;
+    using ComplexT = typename TNDevice_T::ComplexT;
+    using PrecisionT = typename TNDevice_T::PrecisionT;
+
+    const int num_qubits = 4;
+    std::size_t maxExtent = 2;
+
+    std::unique_ptr<TNDevice_T> tn_state0;
+    std::unique_ptr<TNDevice_T> tn_state1;
+
+    if constexpr (std::is_same_v<TestType, MPSTNCuda<double>> ||
+                  std::is_same_v<TestType, MPSTNCuda<float>>) {
+        // Create the object for MPSTNCuda
+        tn_state0 = std::make_unique<TNDevice_T>(num_qubits, maxExtent);
+        tn_state1 = std::make_unique<TNDevice_T>(num_qubits, maxExtent);
+    } else {
+        // Create the object for ExaTNCuda
+        tn_state0 = std::make_unique<TNDevice_T>(num_qubits);
+        tn_state1 = std::make_unique<TNDevice_T>(num_qubits);
+    }
+
+    const auto margin = PrecisionT{1e-5};
+    const std::size_t control = GENERATE(0, 1, 2, 3);
+    const std::size_t wire = GENERATE(0, 1, 2, 3);
+
+    DYNAMIC_SECTION("Controlled gates with base operation - "
+                    << "controls = {" << control << "} "
+                    << ", wires = {" << wire << "} - "
+                    << PrecisionToName<PrecisionT>::value) {
+        if (control != wire) {
+            tn_state0->applyControlledOperation(
+                "PauliX", std::vector<std::size_t>{control},
+                std::vector<bool>{true}, std::vector<std::size_t>{wire});
+
+            tn_state1->applyOperation(
+                "CNOT", std::vector<std::size_t>{control, wire}, false);
+
+            REQUIRE(tn_state0->getDataVector() ==
+                    approx(tn_state1->getDataVector()).margin(margin));
+        }
+    }
+
+    DYNAMIC_SECTION("Controlled gates with a target matrix - "
+                    << "controls = {" << control << "} "
+                    << ", wires = {" << wire << "} - "
+                    << PrecisionToName<PrecisionT>::value) {
+        if (control != wire) {
+            std::vector<ComplexT> gate_matrix = {
+                ComplexT{0.0, 0.0}, ComplexT{1.0, 0.0}, ComplexT{1.0, 0.0},
+                ComplexT{0.0, 0.0}};
+            tn_state0->applyControlledOperation(
+                "applyControlledGates", std::vector<std::size_t>{control},
+                std::vector<bool>{true}, std::vector<std::size_t>{wire}, false,
+                {}, gate_matrix);
+
+            tn_state1->applyOperation(
+                "CNOT", std::vector<std::size_t>{control, wire}, false);
+
+            REQUIRE(tn_state0->getDataVector() ==
+                    approx(tn_state1->getDataVector()).margin(margin));
+        }
+    }
+}
+
+TEMPLATE_LIST_TEST_CASE("TNCuda::applyMPO::2+_wires", "[TNCuda_Nonparam]",
+                        TestTNBackends) {
+    const bool inverse = GENERATE(false, true);
```

Comment:
Do we want to loop both mps and exatn?

Reply:
but exatn does not support MPO right?

Reply:
Ok I will create the new test file for MPO